### PR TITLE
Fix: PUT message properties crash

### PR DESCRIPTION
**Pull request by 678098**

### CODE REVIEW

File: src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp [hunk 1]
```diff
@@ -613,6 +621,9 @@ int PutMessageIterator::next()
         // will use 'header()' as their input (not the blob).
     }
 
+    bmqt::CompressionAlgorithmType::Enum cat =
+        d_header.compressionAlgorithmType();
```
Comment (678098) [inline src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp:625]: Moved this variable closer to the place where it is used

File: src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp [hunk 2]
```diff
@@ -581,13 +582,20 @@ int PutMessageIterator::next()
...
     const bool haveNewMPs = MessagePropertiesInfo::hasSchema(d_header);
-    int        length     = compressedApplicationDataSize();
+
+    // Validation: if new message properties schema is set, the flag
+    // e_MESSAGE_PROPERTIES must also be set.
+    if (BSLS_PERFORMANCEHINT_PREDICT_UNLIKELY(!haveMPs && haveNewMPs)) {
+        BSLS_PERFORMANCEHINT_UNLIKELY_HINT;
+        d_advanceLength = -1;
+        return rc_INVALID_MESSAGE_PROPERTIES;  // RETURN
+    }
```
Comment (678098) [inline src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp:596]: Fail if this is true: `!haveMPs && haveNewMPs`
Let's revert this expression and simplify it:
`!(!haveMPs && haveNewMPs)` ->
`(!!haveMPs) || (!haveNewMPs)` ->
`haveMPs || (!haveNewMPs)` ->
`haveMPs || (!(wire.value() != k_NO_WIRE_SCHEMA))` ->
`haveMPs || (wire.value() == k_NO_WIRE_SCHEMA)` ->
`haveMPs || wire.value() == k_NO_WIRE_SCHEMA`

So, the reverted expression here is equivalent to the offending failing assertion:
`BSLS_ASSERT_SAFE(d_isPresent || d_schemaWireId == k_NO_WIRE_SCHEMA);`

**Review by dorjesinpo — APPROVED** [review]
two very minor comments.  feel free to ignore

File: src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp [hunk 3]
```diff
@@ -607,12 +615,10 @@ int PutMessageIterator::next()
         // (with lengths, not offsets).  That style does not work with schema.
         // So, leave the schema absence. The only effect is de-compressing.
         decompressFlag = true;
-        d_header.setCompressionAlgorithmType(
-            bmqt::CompressionAlgorithmType::e_NONE);
-        // No need to write the header back to the blob since future events
-        // will use 'header()' as their input (not the blob).
```
Comment (678098) [inline src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp:613]: @dorjesinpo I moved this call to the bottom.
Before this PR, we had a following sequence of actions for decompressing old message properties:
`
save algorithm type to 'cat'
set decompress flag
set algorithm type to NONE
decompress using 'cat'
recalculate crc32
`

I changed the order to:
`
set decompress flag
save algorithm type to 'cat'
decompress using 'cat'
set algorithm type to NONE
recalculate crc32
`

Comment (dorjesinpo) [reply src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp:613]: Yeah, we need to read `cat` before  `d_header.setCompressionAlgorithmType(bmqt::CompressionAlgorithmType::e_NONE);`

**Review by dorjesinpo — APPROVED** [review]

**Pull request merged**